Terminate queries

boggle opened the pull request:
Add a new method to KernelTransactions that helps listing all transactions that are registered to execute a given cypher query (ExecutingQuery) and introduces new procedures that use this method to kill the transactions for some given queries.

Also includes some additional cosmetic changes:
- queryIds are now strings on the surface (i.e. what comes back/goes into procedures)
- moves transaction/connection/query management procedures over to kernel
- adds procedure descriptions 

changelog [procedures] Added procedures for listing and killing currently running queries

boggle committed:
Change transaction and connection management procedures to be built-ins

boggle committed:
Add procedure for terminating queries

boggle committed:
Add test to ensure we have an auth subject even if not authenticated in core API

boggle committed:
Change query ids to be strings on the surface

boggle committed:
Rename terminate query to kill query

boggle committed:
Add support for killing multiple queries with a single procedure call

boggle committed:
Add descriptions for list and kill query procedures

boggle committed:
Add testing of built in procedure support to ha and core|edge edition modules

henriknyman committed:
Fix using enterprise built-in procedures over REST

Commit:
Checkstyle fixes

OliviaYtterbrink commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,263 @@
+            throws InvalidArgumentsException
+    {
+        EnterpriseAuthSubject subject = EnterpriseAuthSubject.castOrFail( authSubject );
+        subject.ensureUserExistsWithName( username );
```
Shouldn't this check be after the check for admin or self? Otherwise we could leak information about what users are available to a non admin user.

OliviaYtterbrink commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,263 @@
+            throw new AuthorizationViolationException( PERMISSION_DENIED );
+        }
+
+        subject.ensureUserExistsWithName( username );
```
Could use `ensureSelfOrAdminEnterpriseAuthSubject` here as well

OliviaYtterbrink commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,263 @@
+    public Stream<QueryTerminationResult> killQuery( @Name( "id" ) String idText )
+            throws InvalidArgumentsException, IOException
+    {
+        long queryId = parseQueryId( idText ).kernelQueryId();
```
Don't we want to check permissions for the user running this procedure?

OliviaYtterbrink commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/QueryId.java:
```diff
@@ -0,0 +1,90 @@
+        }
+        catch ( NumberFormatException e )
+        {
+            throw new IllegalArgumentException( e );
```
Should use `InvalidArgumentsException` instead since that has status.

boggle replied:
No, that exception is tied to the security package (same for status). Don't think that's appropriate to use here.

fickludd replied:
I think we can move the `InvalidArgumentsException` and `Status.Security.InvalidArguments` to somewhere more appropriate then. They really have nothing in particular to do with security. Getting an exception with a status gives much nicer handling in driver/browser/logging.

MishaDemianenko replied:
This exception missing error details and expected format. It would be strange to get illegal argument exception without explanation what format is actually expected from you

boggle replied:
Moved InvalidArguments out of security and used it. This requires though catching it and rethrowing from all use in stream pipelines :/

fickludd commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/StandardBoltConnectionTracker.java:
```diff
@@ -53,7 +53,11 @@ public void onTerminate( ManagedBoltStateMachine machine )
+                .entrySet()
+                .stream()
+                .filter( entry -> entry.getValue()
+                .equals( owner ) )
```
This line should probably join it's buddy above, in lack of anything important to say...

boggle replied:
Indeed, let's not keep them lonely :)

fickludd commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java:
```diff
@@ -362,12 +276,13 @@ private StandardEnterpriseAuthSubject ensureAdminAuthSubject()
         subject.getUserManager().getUser( username );
 
-        if ( subject.isAdmin() || subject.doesUsernameMatch( username ) )
+        if ( subject.isAdmin() || subject.hasUsername( username ) )
```
This should be reordered as Olivia said above. I guess
`if ( subject.hasUsername( username ) || 
     ( subject.isAdmin() && subject.getUserManager().getUser( username ) ) )` would cut it

henriknyman commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,263 @@
+        Set<ExecutingQuery> executingQueries = getKernelTransactions().executingQueries();
+        return executingQueries
+            .stream()
+            .filter( query -> isAdminEnterpriseAuthSubject() || authSubject.hasUsername( query.authSubjectName() ) )
```
query.authSubjectName() is using name() which is not the same as username()

henriknyman commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,263 @@
+    private QueryTerminationResult killQueryTransaction( Pair<KernelTransactionHandle, ExecutingQuery> pair )
+    {
+        ExecutingQuery query = pair.other();
+        if ( isAdminEnterpriseAuthSubject() || authSubject.hasUsername( query.authSubjectName() ) )
```
query.authSubjectName() is using name() which is not the same as username()

boggle committed:
Address feedback from PR review

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AuthSubject.java:
```diff
@@ -54,6 +54,22 @@
+     * @param username a username
+     * @throws InvalidArgumentsException if the provided user name is not the name of an existing user
+     */
+    default void ensureUserExistsWithName( String username ) throws InvalidArgumentsException {
```
I find 2 strange things with thin method here: 
- i think ensureUserExistsWithName should not be part of AuthSubject interface at all since it does not look like that AuthSubject should be responsible for creation of users
- since it's our first implementation of AuthSubject interface i believe that we should not declare it `default`. If we need some default behaviour it should be an abstract class instead

boggle replied:
This is unrelated to this PR. We're just moving code around here, I'll try to factor our the default implementation. This whole hierarchy needs some overhaul but out of scope for this PR.

boggle replied:
I think factoring this out just complicates things uneccesarliy and having the default implementation generate the correct errors seems reasonable to me so I'm not going to touch it.

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java:
```diff
@@ -54,56 +70,258 @@
+
+    private Stream<TransactionTerminationResult> terminateTransactionsForValidUser( String username )
+    {
+        long terminatedCount = 0;
```
hm looks like a place where stream api should be used to count terminatedCount instead of for loop

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -30,16 +30,21 @@
     {
         this.queryId = queryId;
-        this.authSubjectName = authSubjectName;
+        this.username = userNameOrNull;
```
silent loss of can be null information, Optional instead?

boggle replied:
Good catch, Optional it is.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -69,14 +74,14 @@ public int hashCode()
     }
 
-    public long queryId()
+    public long kernelQueryId()
```
silent conversion from query id to kernelQueryId

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java:
```diff
@@ -201,9 +203,10 @@ final void forceClose()
+    final String username()
     {
-        return transaction.mode().name();
+        AccessMode mode = transaction.mode();
```
i find it strange that kernel statement perform this dance, when we will have mode not an instance of AuthSubject?

boggle replied:
Always, AuthSubject extends AccessMode. This is in need of refactoring and hopefully that hopefully will happen when security support is added to core api. For now, we need this dance and this situation will arise in community.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -202,6 +205,24 @@ protected void dispose( KernelTransactionImplementation tx )
+     *
+     * @return the (approximate) set of open transactions.
+     */
+    public <T> Set<Pair<KernelTransactionHandle, T>> activeTransactions(
```
do not think this method should be introduced in KernelTransaction at all:
 it's copy paste of active transactions method, it should not return too generic pair thingy, used only in build in procedures so mapping, if its necessary can be produced there

boggle replied:
I'll pull it up.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/query/Neo4jTransactionalContextTest.java:
```diff
@@ -154,4 +154,78 @@ private void setUpMocks()
+    public void rollsBackNewlyCreatedTransactionIfTerminationDetectedOnCloseDuringPeriodicCommit()
+    {
+        // Given
+        GraphDatabaseQueryService graph = mock( GraphDatabaseQueryService.class );
```
i find this test extremely unreadable and unsupportable, only one thing you will do when it will fail - make mocks happy

boggle replied:
That would be bad practice. If this fails, we should stop and think why.  I understand that this is not perfect yet it is the only good test we felt we could write here. We didn't find a better way to test the behavior we need to rely on in the built in procedures without resorting to heavy use of mocking or probabilistic (yuck) testing. How would you test this?

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/QueryId.java:
```diff
@@ -0,0 +1,90 @@
+ */
+package org.neo4j.kernel.enterprise.builtinprocs;
+
+public final class QueryId
```
Why do we need to represent query externally as query-_queryId_ ? would it be more simpler just to have an numeric id?

boggle replied:
we want to keep the door open for cluster wide ids and we're not sure longs will be good enough for that.

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/QueryId.java:
```diff
@@ -0,0 +1,90 @@
+    public static final String QUERY_ID_PREFIX = "query-";
+    private final long kernelQueryId;
+
+    public static QueryId queryId( long queryId )
```
i believe static factory methods here should be called _of_

MishaDemianenko commented:
@boggle @systay i finish my first round of review

boggle committed:
Use InvalidArgumentsException in BuiltInProcedures

o InvalidArgumentsException is now a generic exception and no longer tied to security
o Wrote some utility code for tunneling checked exceptions out of stream pipelines

boggle committed:
Use Optional<String> to represent usernames for query registration purposes

boggle committed:
Refactor according to PR feedback

boggle commented:
@MishaDemianenko addressed feedback

MishaDemianenko commented:
@boggle @systay do not have kernel questions anymore feel free to merge on green! thx

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -98,7 +104,7 @@ public long startTime()
-                "ExecutingQuery{queryId=%d, authSubjectName='%s', queryText='%s', queryParameters=%s, startTime=%d}",
-                queryId, authSubjectName, queryText, queryParameters, startTime );
+            "ExecutingQuery{queryId=%d, username='%s', queryText='%s', queryParameters=%s, startTime=%d}",
+            queryId, username == null ? "<unavailable>" : username, queryText, queryParameters, startTime );
```
Not using username as optional

OliviaYtterbrink commented on enterprise/backup/data/databases/to/logs/debug.log:
```diff
@@ -0,0 +1,598 @@
+2016-09-12 14:11:52.756+0000 INFO  [o.n.k.i.f.CommunityEditionModule] No locking implementation specified, defaulting to 'forseti'
```
I don't believe these files should be here... ;p

boggle committed:
Improved handling of unavailable usernames

boggle committed:
Add notorious enterprise/backup/data to .gitignore

Commit:
Fix compilation problem

boggle committed:
Fix more compilation problems
